# 1th_hometask

Всем привет!

Меня вы уже знаете, я ваш преподаватель, Артур Сапрыкин 🙂 

Напомню то, о чём говорили про домашки:
1. Домашки задаю после каждого занятия
2. Сдавать их нужно до начала занятия вторника следующей недели. То есть, если дз задано во вторник и четверг на одной неделе, то на следующей, перед новым блоком, нужно всё сдать.
3. Если не успели сдать домашку, то проверять её не буду. НО вы всегда можете для себя её сделать 😉
4. И очень важное - нельзя выкладывать результаты работы в чат и обмениваться решениями с другими участниками до начала следующего модуля!


Это ваша первая домашка.

Вам нужно сделать предобработку данных, аналогичную на занятии. 

Расписывайте в Markdown ваши шаги, и почему вы взяли медину вместо среднего, например.

Это покажет то, как вы рассуждаете.

Домашка простая. Делать немного.

Результаты можете слать в личку или ссылки на гитхаб.

Всем успехов!

### Import Section

In [1]:
import numpy as np
import pandas as pd

### Function "Reduce Memory Usage"

In [2]:
def reduce_memory_usage(df):
    
    # СДЕЛАТЬ ПРОВЕРКУ type(df) == pd.DataFrame
    
    initial_memory_usage = df.memory_usage().sum() / 1024 / 1024
    print(f'Initial memory usage of dataframe:\t{initial_memory_usage:.3} Mb')
    
    for col in df.columns:
        col_type = df[col].dtype
        
        if col_type != 'object':
            c_min = df[col].min()
            c_max = df[col].max()
            
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)
            
            # Можно было бы сделать оптимизацию вплоть до типа "float16", однако данный тип данных,
            # как сообщается от сообщества аналитиков данных, плохо поддерживается некоторыми библиотеками
            
            if str(col_type)[:5] == 'float':
                if c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                elif c_min > np.finfo(np.float64).min and c_max < np.finfo(np.float64).max:
                    df[col] = df[col].astype(np.float64)
        
        else:
            df[col] = df[col].astype('category')
    
    final_memory_usage = df.memory_usage().sum() / 1024 / 1024
    print(f'Final memory usage of dataframe:\t{final_memory_usage:.3} Mb')
    
    comparison = np.round(100 * (initial_memory_usage - final_memory_usage) / initial_memory_usage, 3)
    print(f'Memory usage has been decreased by:\t{comparison} %')
    
    return df

### Path Section

In [197]:
PATH_DATA = r'data.csv'
PATH_DATA_PREPROCESSED = r'data_preprocessed.csv'

## Exploratory Data Analysis

In [4]:
df_data = pd.read_csv(PATH_DATA)

In [5]:
df_data.head(6)

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   
5   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   
5  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                                Event Medal  
0         Basketball Men's Basketball   NaN  
1        Judo Men's Extra-Lightweight   NaN  
2             Football Men's Football   NaN  
3         Tug-Of-War Men's Tug-Of-War  Gold  
4    Speed Skating Women's 500 metres   NaN  
5  Speed Skating Women's 1,000 metres   NaN

In [6]:
df_data.describe()

ID            Age         Height         Weight  \
count  271116.000000  261642.000000  210945.000000  208241.000000   
mean    68248.954396      25.556898     175.338970      70.702393   
std     39022.286345       6.393561      10.518462      14.348020   
min         1.000000      10.000000     127.000000      25.000000   
25%     34643.000000      21.000000     168.000000      60.000000   
50%     68205.000000      24.000000     175.000000      70.000000   
75%    102097.250000      28.000000     183.000000      79.000000   
max    135571.000000      97.000000     226.000000     214.000000   

                Year  
count  271116.000000  
mean     1978.378480  
std        29.877632  
min      1896.000000  
25%      1960.000000  
50%      1988.000000  
75%      2002.000000  
max      2016.000000

##### Age: содержит выбросы/ошибки. Минимальный возраст 10, максимальный - 97, что маловероятно.
##### Height: содержит выбросы/ошибки. Минимальный рост 127, что маловероятно.
##### Weight: содержит выбросы/ошибки. Минимальный вес 25, что маловероятно.
##### Year: необходимо дополнительно рассмотреть значения в районе 1-вый перцентили, возможны ошибки.

In [7]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
dtypes: float64(3), int64(2), object(10)
memory usage: 31.0+ MB


In [39]:
df_null = pd.DataFrame(df_data.isna().sum(), columns=['Nulls'])
df_null = df_null[df_null['Nulls'] > 0]
df_null['Total'] = df_data.shape[0]
df_null['Percentage'] = df_null['Nulls'] / df_null['Total'] * 100
df_null

Nulls   Total  Percentage
Age       9474  271116    3.494445
Height   60171  271116   22.193821
Weight   62875  271116   23.191180
Medal   231333  271116   85.326207

##### Изначальный набор данных содержит значительное количество пропусков, более 22%, среди важных признаков Height и Weight. Данные пропуски не могут быть исключены из набора ввиду их существенной доли и должны быть обработы.

##### Также присутствует незначительное количество пропусков признака Age, около 3,5% записей. Данные пропуски могут как обработаны, так и исключены из набора.

##### Признак Medal является целевой переменной. Примем предположение, что отсуствие информации о медале - отсутствие призового места.

### ID и Name
##### Изначально планировалось изпользовать признак "ID" в качестве индекса, однако данный признак не является уникальным.

In [158]:
df_id_counts = pd.DataFrame(df_data['ID'].value_counts())
df_id_counts[df_id_counts['ID'] > 1].head()

ID
77710   58
106296  39
115354  38
119591  36
129196  32

##### Рассмотрим подробнее записи с одинаковыми id.

In [9]:
non_unique_ids = df_id_counts[df_id_counts['ID'] > 1].index
non_unique_ids

Int64Index([ 77710, 106296, 115354, 119591, 129196,  44875,  53240, 119590,
             89187, 106156,
            ...
             32127, 111077,  32338, 111186, 117178, 117011,  42204,  31547,
             32349,  32093],
           dtype='int64', length=57670)

In [10]:
df_data[df_data['ID'] == non_unique_ids[0]].head()

ID                  Name Sex   Age  Height  Weight    Team  NOC  \
154798  77710  Robert Tait McKenzie   M  44.0     NaN     NaN  Canada  CAN   
154799  77710  Robert Tait McKenzie   M  60.0     NaN     NaN  Canada  CAN   
154800  77710  Robert Tait McKenzie   M  60.0     NaN     NaN  Canada  CAN   
154801  77710  Robert Tait McKenzie   M  60.0     NaN     NaN  Canada  CAN   
154802  77710  Robert Tait McKenzie   M  60.0     NaN     NaN  Canada  CAN   

              Games  Year  Season       City             Sport  \
154798  1912 Summer  1912  Summer  Stockholm  Art Competitions   
154799  1928 Summer  1928  Summer  Amsterdam  Art Competitions   
154800  1928 Summer  1928  Summer  Amsterdam  Art Competitions   
154801  1928 Summer  1928  Summer  Amsterdam  Art Competitions   
154802  1928 Summer  1928  Summer  Amsterdam  Art Competitions   

                                              Event Medal  
154798           Art Competitions Mixed Sculpturing   NaN  
154799  Art Competitions Mixed Sculpturing, Statues   NaN  
154800  Art Competitions Mixed Sculpturing, Statues   NaN  
154801  Art Competitions Mixed Sculpturing, Statues   NaN  
154802  Art Competitions Mixed Sculpturing, Statues   NaN

In [11]:
df_data[df_data['ID'] == non_unique_ids[1]].head()

ID                      Name Sex   Age  Height  Weight     Team  \
211668  106296  Heikki Ilmari Savolainen   M  20.0   172.0    64.0  Finland   
211669  106296  Heikki Ilmari Savolainen   M  20.0   172.0    64.0  Finland   
211670  106296  Heikki Ilmari Savolainen   M  20.0   172.0    64.0  Finland   
211671  106296  Heikki Ilmari Savolainen   M  20.0   172.0    64.0  Finland   
211672  106296  Heikki Ilmari Savolainen   M  20.0   172.0    64.0  Finland   

        NOC        Games  Year  Season       City       Sport  \
211668  FIN  1928 Summer  1928  Summer  Amsterdam  Gymnastics   
211669  FIN  1928 Summer  1928  Summer  Amsterdam  Gymnastics   
211670  FIN  1928 Summer  1928  Summer  Amsterdam  Gymnastics   
211671  FIN  1928 Summer  1928  Summer  Amsterdam  Gymnastics   
211672  FIN  1928 Summer  1928  Summer  Amsterdam  Gymnastics   

                                         Event Medal  
211668  Gymnastics Men's Individual All-Around   NaN  
211669        Gymnastics Men's Team All-Around   NaN  
211670            Gymnastics Men's Horse Vault   NaN  
211671          Gymnastics Men's Parallel Bars   NaN  
211672         Gymnastics Men's Horizontal Bar   NaN

##### Очевидно, что признак ID является уникальным идентификатором каждого спортсмена. Теоретически, может возникнуть ситуация, когда признак ID позволит однозначно идентифицировать спортсменов с одинаковыми именами, полом и другими характеристиками.

##### Дальнейшая задача прогнозирования разбивается на две подзадачи:

##### 1. Прогнозирование для каждого конкретного спортсмена (Когда имя спортсмена оказывает влияние на результат, например опыт конкретного спортсмена или травма). В таком случае наиболее важен уникальный ID каждого спортсмена и признаки, описывающие условия (вид спорта, вид соревнования, событие), в которых конкретный спортсмен выступает.

##### 2. Прогнозирование исходя из характеристик спортсменов. В таком случае спортсменов необходимо рассматривать анонимно, без влияния их ID и Name на результат. В данном случае будут важны и характеристики спортсменов (рост, вес, пол, происхождение) и условия (вид спорта, вид соревнования, событие).

##### Далее будем рассматривать подзадачу 2, так как подзадача 1 плохо вписывается в область ML (привязка к конкретным спортсменам).

##### p.s. Согласно правилам реляционных отношений (SQL, DDL) данный набор данных можно было бы разбить на несколько таблиц в целях оптимизации.

### ID и Name вывод:
##### Признаки ID и Name не принесут статистически важной информации для прогнозирования в случае, когда рассматриваются только характеристики спортсменов.

### Sex

##### Бинарный признак, может быть перекодирован.

### Age, Height, Weight

##### Данные признаки содержат пропуски, а также они наиболее важны для прогнозирования целевой переменной Medal.

##### Признаки Height и Weight можно было бы трансформировать в Индекс Массы Тела, но в таком случае возможна потеря информации.

##### Как выяснилось ранее, признак ID является уникальным идентификатором спортсмена, что может быть использовано для заполнения пропусков. Информация для заполнения пропусков может быть получена из других записей с идентичным ID.

In [157]:
df_nulls = df_data[['ID', 'Age', 'Height', 'Weight']]
df_id_counts = pd.DataFrame(df_nulls['ID'].value_counts()).reset_index().rename({'ID': 'ID_counts'}, axis=1).rename({'index': 'ID'}, axis=1)
df_nulls = df_nulls.merge(df_id_counts, how='left', on='ID')

df_age_counts = pd.DataFrame(df_nulls['ID'][df_nulls['Age'].isna()].value_counts()).reset_index().rename({'ID': 'Age_null_counts'}, axis=1).rename({'index': 'ID'}, axis=1)
df_nulls = df_nulls.merge(df_age_counts, how='left', on='ID')

df_height_counts = pd.DataFrame(df_nulls['ID'][df_nulls['Height'].isna()].value_counts()).reset_index().rename({'ID': 'Height_null_counts'}, axis=1).rename({'index': 'ID'}, axis=1)
df_nulls = df_nulls.merge(df_height_counts, how='left', on='ID')

df_weight_counts = pd.DataFrame(df_nulls['ID'][df_nulls['Weight'].isna()].value_counts()).reset_index().rename({'ID': 'Weight_null_counts'}, axis=1).rename({'index': 'ID'}, axis=1)
df_nulls = df_nulls.merge(df_weight_counts, how='left', on='ID')

df_nulls = df_nulls.drop_duplicates()
df_nulls[(df_nulls['ID_counts'] > df_nulls['Age_null_counts']) | (df_nulls['ID_counts'] > df_nulls['Height_null_counts']) | (df_nulls['ID_counts'] > df_nulls['Weight_null_counts'])].head()

Empty DataFrame
Columns: [ID, Age, Height, Weight, ID_counts, Age_null_counts, Height_null_counts, Weight_null_counts]
Index: []

### Age, Height, Weight вывод:

##### Не обнаружилось ни одной записи, где пропуски можно было бы заполнить из другой записи с идентичным ID. Это значит, что информация о некоторых спортсменах отсутствует полностью и должна быть сгенерирована.

##### Биологические характеристики стремятся к нормальному распределению, однако в случае недостаточно большой выборки из генеральной совокупности, может наблюдаться дисбаланс. Чтобы избежать влияния выбросов и дисбаланса используем медиану для заполнения пропусков.

##### Также можно было бы применить обработку с разбиением по регионам. Например, для спортсменов из Канады использовать медиану только канадских спортсменов.

##### Также необходима обработка аномалий.

### Team и NOC

##### Данные признаки содержат практически идентичную информацию, так как в Team записана страна, а в NOC - National Occupational Classification, что отражает тоже самое.

##### Таким образом, можно оставить только один признак. Пусть это будет NOC как официальный международный классификатор.

### Games, Year, Season

##### Данные признаки также копируют информацию друг друга. Ввиду того, что удобнее работать с Year и Season отдельно, от признака Games можно избавиться. Признак Season может быть переведён в категориальный.

##### Признак Year содержит значения < 1900, что может свидетельствовать об ошибке. Однако установить факт ошибки не представляется возможным.

### City, Event

##### Теоретически, данные признаки могут оказывать влияние на целевую переменную, однако в случае честных соревнований и объективного непредвзятого судейства все условия должны быть равны для спортсменов, независимо от места проведения события и самого события. Таким образом, данные признаки могут быть удалены в данный момент, но иметь потенциал для дальнейшего усовершенствования модели в дальнейшем.

### Sport

##### Очевидно, что данный признак должен оказывать влияние на целевую переменную. Например в случае, когда один спортсмен участвует в разных видах спорта и имеет разные преимущества и недостатки в разных видах спорта. Данный признак может быть переведён в категориальный.

### Medal

##### Целевая переменная. Должна быть переведена в категориальный тип, так как нельзя однозначно сказать, какое числовое соотношение будут иметь золотые, серебрянные и бронзовые медали.

## Data Preprocessing

##### Исчклюсения признаков, заполнение пропусков, обработка аномалий (заполнение пертентилями).

In [193]:
df_data_preprocessed = df_data.copy()

df_data_preprocessed = df_data_preprocessed.drop(columns=['ID', 'Name', 'Team', 'Games', 'City', 'Event'])

df_data_preprocessed['Sex_M'] = False
df_data_preprocessed['Sex_M'][df_data_preprocessed['Sex'] == 'M'] = True
df_data_preprocessed = df_data_preprocessed.drop(columns=['Sex'])

df_data_preprocessed['Medal_Gold'] = False
df_data_preprocessed['Medal_Silver'] = False
df_data_preprocessed['Medal_Bronze'] = False
df_data_preprocessed['Medal_Gold'][df_data_preprocessed['Medal'] == 'Gold'] = True
df_data_preprocessed['Medal_Silver'][df_data_preprocessed['Medal'] == 'Silver'] = True
df_data_preprocessed['Medal_Bronze'][df_data_preprocessed['Medal'] == 'Bronze'] = True
df_data_preprocessed = df_data_preprocessed.drop(columns=['Medal'])


df_data_preprocessed['Age'] = df_data_preprocessed['Age'].fillna(value=df_data_preprocessed['Age'].median())
df_data_preprocessed['Age'][df_data_preprocessed['Age'] > df_data_preprocessed['Age'].quantile(q=0.999)] = df_data_preprocessed['Age'].quantile(q=0.999)

df_data_preprocessed['Height'] = df_data_preprocessed['Height'].fillna(value=df_data_preprocessed['Height'].median())

df_data_preprocessed['Weight'] = df_data_preprocessed['Weight'].fillna(value=df_data_preprocessed['Weight'].median())
df_data_preprocessed['Weight'][df_data_preprocessed['Weight'] > df_data_preprocessed['Weight'].quantile(q=0.95)] = df_data_preprocessed['Weight'].quantile(q=0.95)

df_data_preprocessed.head()

C:\Users\SERGEY~1\AppData\Local\Temp/ipykernel_5724/381600026.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_data_preprocessed['Sex_M'][df_data_preprocessed['Sex'] == 'M'] = True
C:\Users\SERGEY~1\AppData\Local\Temp/ipykernel_5724/381600026.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_data_preprocessed['Medal_Gold'][df_data_preprocessed['Medal'] == 'Gold'] = True
C:\Users\SERGEY~1\AppData\Local\Temp/ipykernel_5724/381600026.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stab

Age  Height  Weight  NOC  Year  Season          Sport  Sex_M  Medal_Gold  \
0  24.0   180.0    80.0  CHN  1992  Summer     Basketball   True       False   
1  23.0   170.0    60.0  CHN  2012  Summer           Judo   True       False   
2  24.0   175.0    70.0  DEN  1920  Summer       Football   True       False   
3  34.0   175.0    70.0  DEN  1900  Summer     Tug-Of-War   True        True   
4  21.0   185.0    82.0  NED  1988  Winter  Speed Skating  False       False   

   Medal_Silver  Medal_Bronze  
0         False         False  
1         False         False  
2         False         False  
3         False         False  
4         False         False

In [194]:
df_data_preprocessed.describe()

Age        Height         Weight           Year
count  271116.000000  271116.00000  271116.000000  271116.000000
mean       25.497171     175.26374      70.028493    1978.378480
std         6.250141       9.27917      11.094528      29.877632
min        10.000000     127.00000      25.000000    1896.000000
25%        22.000000     170.00000      63.000000    1960.000000
50%        24.000000     175.00000      70.000000    1988.000000
75%        28.000000     180.00000      75.000000    2002.000000
max        65.000000     226.00000      93.000000    2016.000000

In [196]:
df_data_preprocessed = reduce_memory_usage(df_data_preprocessed)

Initial memory usage of dataframe:	15.5 Mb
Final memory usage of dataframe:	5.7 Mb
Memory usage has been decreased by:	63.254 %


In [198]:
df_data_preprocessed.to_csv(PATH_DATA_PREPROCESSED)